File: SecondYear/semestrul4/MPP/PROIECT/CSharpProjectNetworking/Persistence/JdbcUtils.cs

using System.Configuration;
using System.Data.Common;
using log4net;
using Microsoft.Data.Sqlite;

namespace Persistence;

public class JdbcUtils
{
    private static readonly ILog logger = LogManager.GetLogger(typeof(JdbcUtils));
    private static JdbcUtils instance;
    private readonly DbConnection connection;

    private JdbcUtils()
    {
        try
        {
            logger.Info("Loading database configuration from app.config...");

            var connectionStringSettings = ConfigurationManager.ConnectionStrings["SQLiteConnection"];

            if (connectionStringSettings == null)
            {
                throw new Exception("No valid database connection string found in app.config.");
            }

            string providerName = connectionStringSettings.ProviderName;
            string connectionString = connectionStringSettings.ConnectionString;

            if (string.IsNullOrEmpty(connectionString))
            {
                throw new Exception("Database connection string is missing in app.config.");
            }

            logger.Info($"Connecting to database using provider: {providerName}");
            
            if (providerName == "System.Data.SQLite")
            {
                connection = new SqliteConnection(connectionString);
            }
            else
            {
                throw new Exception($"Unsupported database provider: {providerName}");
            }

            connection.Open();
            logger.Info("Database connection established successfully.");
        }
        catch (Exception ex)
        {
            logger.Error("Database connection failed.", ex);
            throw new Exception("Database connection failed.", ex);
        }
    }

    public static JdbcUtils GetInstance()
    {
        return instance ??= new JdbcUtils();
    }

    public DbConnection GetConnection()
    {
        logger.Info("Getting database connection...");
        return connection;
    }
}

File: SecondYear/semestrul4/MPP/examen_practic/Rezolvari/generic/persistence/src/main/java/persistence/utils/JdbcUtils.java
package persistence.utils;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Properties;

public class JdbcUtils {


    private static final Logger logger= LogManager.getLogger(JdbcUtils.class);
    private static JdbcUtils instance;
    private final Connection connection;

    public JdbcUtils(){
        try {
            logger.info("Loading database configuration...");
            Properties config = PropertyFileLoader.loadProperties("/Users/croitoruandreea/Desktop/ANUL2/semestrul4/MPP/examen_practic/boat/boat/bd.config");
            String URL = config.getProperty("jdbc.url");
            String USER = config.getProperty("jdbc.user");
            String PASS = config.getProperty("jdbc.pass");

            logger.info("Attempting to connect to database...");
            connection = DriverManager.getConnection(URL, USER, PASS);

            logger.info("Database connection established successfully.");
        } catch (SQLException e) {
            logger.error("Database connection failed.", e);
            throw new RuntimeException(e);
        }
    }

    public static JdbcUtils getInstance() {
        if (instance == null) {
            instance = new JdbcUtils();
        }
        return instance;
    }

    public Connection getConnection(){
        logger.traceEntry();
        try{
            return connection;
        }catch (Exception e){
            logger.error("Connection lost", e);
            throw new RuntimeException(e);
        }


    }
}
